Support qGetTLSAddr

tromey opened the pull request:
Fixes #146.

This adds support for the qGetTLSAddr and qSymbol packets from gdb,
which lets gdb access TLS in a replay.

rocallahan commented on CONTRIBUTING.md:
`````diff
@@ -22,7 +22,7 @@ Please make sure you go through this list before submitting a patch.  The rules
 - [ ] Did you check your code is formatted correctly? It's easiest to run
 ````
-find src include -name '*.cc' -or -name '*.h' -or -name '*.c'|xargs clang-format -i -style=file
+find src include -name '*.cc' -or -name '*.h' -or -name '*.c'|fgrep -v proc_service.h|xargs clang-format -i -style=file
`````
`proc_service.h` should be under `third_party`. Then this won't be needed.

rocallahan replied:
... in its own subdirectory, with a README explaining where you copied it and what must be done to comply with its license.

tromey replied:
After talking with a more license-interested friend I became pretty confused about what the LGPL really requires here.  I didn't know what to write.  So instead I grabbed the FreeBSD header and made the trivial modifications necessary to make it match glibc.  Wondering if that is ok.

rocallahan replied:
That sounds OK. Thanks.

rocallahan commented on src/TaskGroup.h:
```diff
@@ -125,6 +128,8 @@ class TaskGroup : public HasTaskSet {
 
   uint32_t serial;
+
+  ThreadDb thread_db_;
```
I think we should make this a `unique_ptr` so that we don't need to include `ThreadDB.h` here and `ThreadDB` will not need to be instantiated during recording, which doesn't need this.

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,246 @@
+  iovec.iov_len = sizeof(*result);
+  iovec.iov_base = result;
+
+  if (ptrace((__ptrace_request)PTRACE_GETREGSET, tid, &iovec, NT_PRSTATUS) ==
```
I think it would be a bit nicer to just get these directly from `Task::regs()`, abstracting away the ptrace usage.

tromey replied:
When writing this originally I couldn't convince myself that `regs()` uses a regset; because I was under the impression (perhaps mistaken) that `PTRACE_GETREGS` and `PTRACE_GETREGSET` were not identical.  I will research a bit more to find out for sure.

tromey replied:
After looking at the kernel source I think it's fine.

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,246 @@
+  rr::Task* task = *h->task_group->task_set().begin();
+  task->read_bytes_helper(uaddr, len, buffer, &ok);
+  LOG(debug) << "ps_pdread " << ok;
+  return ok ? PS_OK : PS_ERR;
```
In modern Linux kernels every task in the task group must share VM, and you're assuming that here. Please make that explicit.

OTOH tasks can share VM without being the same task group but I guess you're not assuming that here.

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,246 @@
+      if (ptrace((__ptrace_request)PTRACE_ARCH_PRCTL, tid, base, ARCH_GET_FS) ==
+          0) {
+        LOG(debug) << "ps_get_thread_area FS OK";
+        return PS_OK;
```
You can get these from Task::regs() on x86-64. Does this even get called (or work) on x86-32? Also this `PTRACE_ARCH_PRCTL` is undocumented... if you must use a syscall here, why not use `arch_prctl`?

tromey replied:
It seems for x86-32 inferiors we have to use `PTRACE_GET_THREAD_AREA`, also apparently undocumented.

I don't think `arch_prctl` can be used as, IIUC, it works on the current process, not some other process.

tromey replied:
For x86 we also have to be sure to load a different libthread_db.  (As an aside let me say that this API is unbelievably terrible and I can't wait for the replacement to land.)  I'm not 100% sure of the best way to do this, but it seems we'd need some way to find which libpthread is being used by the inferior, then find the corresponding libthread_db.  (I was under the impression that the 64 bit libthread_db could also understand 32 bit inferiors, but that doesn't seem to be the case.)

rocallahan replied:
> I don't think arch_prctl can be used as, IIUC, it works on the current process, not some other process.

Of course, right.

tromey replied:
I found out that the 64-bit libthread_db does indeed support 32-bit inferiors; and previously my code was wrong.

Also I found out that modern kernels provide fs_base and gs_base in the user regs, so we can fetch those there.

Still not sure whether rr already has the thread area.

tromey replied:
> Still not sure whether rr already has the thread area.

Ok, found that too.  I'll upload the updated patch momentarily.

rocallahan commented:
A heroic effort!

I still need to look a bit closer to grok how the symbol lookups work. A summary comment somewhere about how this all fits together might help.

tromey commented:
> A summary comment somewhere about how this all fits together might help.

I'll fix this and address the other comments (or ask questions...) soon.

The qSymbol handling is kind of explained in the GdbServer.cc patch, but this could be expanded.

Keno commented:
As another point of reference, lldb reads the relevant symbols directly: https://github.com/llvm-mirror/lldb/blob/master/source/Plugins/DynamicLoader/POSIX-DYLD/DYLDRendezvous.cpp#L614

rocallahan commented:
> Is linking against libthread_db really the best approach here?

A good question...

If there's a simple approach that only works for modern glibcs, I think that would be fine.

tromey commented:
I don't think it's that easy, and I've already written this approach, so barring unforseen circumstances I plan to press on.  If someone else writes a better patch, I'd be in favor of that as well.

rocallahan commented:
OK.

Keno commented:
I'll happily spend a couple hours on this. I don't particularly care about GDB, but my front-end doesn't have TLS support yet, so why not take the chance and do it generically.

rocallahan commented on src/GdbConnection.h:
```diff
@@ -197,7 +204,14 @@ struct GdbRequest {
+    remote_ptr<void> load_module;
+  } tls_;
+  struct Symbol {
+    std::string address;
```
I'd prefer to have the address parsed in `GdbConnection`.

tromey replied:
I did it this way so that GdbServer can easily see when there is no address.  Using a sentinel address is possible but seemed less clean.  I suppose I could stick a bool in there as well.

rocallahan commented on src/Session.cc:
```diff
@@ -153,6 +153,16 @@ Task* Session::find_task(const TaskUid& tuid) const {
   return t && t->tuid() == tuid ? t : nullptr;
 }
 
+Task* Session::find_task_tid(pid_t tid) const {
```
Call this `find_task_by_real_tid`

rocallahan commented on src/Task.h:
```diff
@@ -617,6 +617,26 @@ class Task {
+
+  /**
+   * Return a set of the names of all the symbols that have not yet
+   * been resolved.
```
Shouldn't we return the names of all symbols, whether or not they have been resolved? In case symbols already resolved need to be re-resolved to a different address?

tromey replied:
Once they're found they don't get relocated.

rocallahan replied:
What if we reverse-execute backward through the loading of libpthread? I think in that case they'll disappear and returning invalid values might be a problem.

rocallahan commented on src/TaskGroup.h:
```diff
@@ -11,11 +11,13 @@
 
 #include "HasTaskSet.h"
 #include "TaskishUid.h"
+#include "ThreadDb.h"
```
Can we remove this #include?

tromey replied:
Done.  Sorry about that, it was mentioned in the previous review and I didn't deal with it.

rocallahan commented on src/ThreadDb.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * The overall approach is that a libthread_db.so is loaded into rr
+ * when this class is initialized (see |load_library|).  This provides
+ * the GdbServer with list of symbols whose addresses might be needed
```
"with a list of symbols"

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,232 @@
+  bool ok = true;
+  uintptr_t uaddr = reinterpret_cast<uintptr_t>(addr);
+  // We need any task associated with the task group.  Here we assume
+  // that all the tasks in the task group share VM.
```
", which is enforced by clone(2)."

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,232 @@
+
+ps_err_e ps_pdwrite(struct ps_prochandle*, psaddr_t, const void*, size_t) {
+  LOG(debug) << "ps_pdwrite";
+  return PS_ERR;
```
I think it's probably better to do
````
FATAL() << "ps_pdwrite not implemented";
````
If we hit this, that'll be easier to disagnose than some more mysterious failure.

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,232 @@
+}
+
+ps_err_e ps_lgetregs(struct ps_prochandle* h, lwpid_t tid, prgregset_t result) {
+  rr::Task* task = h->task_group->session()->find_task_tid(tid);
```
Hmm. Why are we looking up a task by real-tid here? This is a bit of a dilemma, potentially; if libthreaddb uses /proc, then it will be wanting real-tids, but if it looks into data structures in the replaying process memory, it will be wanting recorded-tids. Have you considered this issue? Are you sure we're OK always using real-tids? If so, please document that clearly somewhere.

rocallahan commented on src/ThreadDb.cc:
```diff
@@ -0,0 +1,232 @@
+
+pid_t ps_getpid(struct ps_prochandle* h) {
+  LOG(debug) << "ps_getpid " << h->task_group->tgid;
+  return h->task_group->tgid;
```
So this is weird, because here you're returning the recorded tgid, so we're being inconsistent...

tromey replied:
libthread_db only calls ps_getpid an unusual situation, I think maybe when it's being used before NPTL has fully initialized.

rocallahan commented:
This is basically looking good. The major issue remaining is this confusion about whether libthreaddb should see recorded-tids or real-tids. My instinct is that it should always see recorded-tids, in which case a few places need to be changed. E.g. the call to `ThreadDb::get_tls_address` from `Task` needs to pass `rec_tid`.

tromey commented:
FWIW I tried the rec_tid approach and backed it out for failures.  However perhaps I didn't try hard enough.  I'll do it again and debug more.

tromey commented:
Changing to rec_tid worked fine this time around, so I probably just missed some detail in the previous attempt.  New patch to follow.  I think I've addressed all the comments.

rocallahan commented:
OK, this is looking good except that I still think when gdb tells us it can re-resolve symbols, I think we should actually re-resolve all symbols.

rocallahan commented:
@tromey would you like me to take this over and get it in?

tromey commented:
@rocallahan - the current PR has all the changes that you requested, as far as I'm aware.  (In particular the one to look up all the symbols each time.)  I suspect maybe you missed the last push... but if there's something more to do, please let me know, and I'll do it on Monday.

rocallahan commented:
Argh! I have email notifications for "Pull Request pushes" checked, but I don't seem to get them :-(. I'll look at this on Monday. Thanks!!!

rocallahan commented:
Excellent. Merged! Thanks so much!

tromey commented:
Thanks!  Please feel free to ping me on any issues that may arise.

Keno commented:
Clang is unhappy:
```
/home/kfischer/rr/src/ThreadDb.h:95:8: error: private field 'initialized' is not used [-Werror,-Wunused-private-field]
  bool initialized;
       ^
```
